### PR TITLE
containers/create: use shareable ipc by default for older clients

### PR DESCRIPTION
**Pull request by kolyshkin**
Older (API < 1.32) clients are not aware of new IPC modes
(`shareable`, `private`, and `none`, introduced in commit 7120976d7419 (PR #34087)
and expect a new container to be created with IPC mode that
is now known as `shareable`.

Since the above commit, daemon default can be set to `private`,
essentially locking out such older clients from being able to
share IPC between containers (i.e. using `--ipc container:NNN`).

So, with this patch, a client which does not set any value
for IPC mode, defaults to:

 * for API < 1.32: `shareable`;
 * for API >= 1.32: daemon's default-ipc-mode

### CODE REVIEW

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
CLI test failing

```
02:44:37 ----------------------------------------------------------------------
02:44:37 FAIL: docker_cli_daemon_test.go:2973: DockerDaemonSuite.TestDaemonRestartIpcMode
02:44:37 
02:44:37 [d6050514c6c8f] waiting for daemon to start
02:44:37 [d6050514c6c8f] daemon started
02:44:37 
02:44:37 docker_cli_daemon_test.go:2990:
02:44:37     c.Assert(m, checker.Equals, "private")
02:44:37 ... obtained string = "shareable"
02:44:37 ... expected string = "private"
02:44:37 
02:44:37 [d6050514c6c8f] exiting daemon
```

**Comment by cpuguy83** [comment]
Let's take a step back here. I don't see why we should modify the client, are we doing something similar for other options?

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by cpuguy83** [comment]
```
19:03:28 FAIL: docker_cli_daemon_test.go:2973: DockerDaemonSuite.TestDaemonRestartIpcMode
19:03:28 
19:03:28 [d77aa0cf0b918] waiting for daemon to start
19:03:28 [d77aa0cf0b918] daemon started
19:03:28 
19:03:28 docker_cli_daemon_test.go:2990:
19:03:28     c.Assert(m, checker.Equals, "private")
19:03:28 ... obtained string = "shareable"
19:03:28 ... expected string = "private"
19:03:28 
19:03:28 [d77aa0cf0b918] exiting daemon
19:03:29 
```

**Comment by kolyshkin** [comment]
> 19:03:28 FAIL: docker_cli_daemon_test.go:2973: DockerDaemonSuite.TestDaemonRestartIpcMode

This test is failing as the API used is < 1.32 (basically confirming the patch works). Fix is coming :)

**Comment by kolyshkin** [comment]
OK here's the test fix; please see individual commits description.

**Comment by kolyshkin** [comment]
1. exp ci failure in 

> 01:05:06.642 FAIL: docker_api_swarm_test.go:297: DockerSwarmSuite.TestAPISwarmLeaderElection

is a known flaky test

2. janky ci failure in

> 01:19:34.071 FAIL: docker_cli_daemon_test.go:118: DockerDaemonSuite.TestDaemonRestartUnlessStopped

is a recent flaky test (https://github.com/moby/moby/issues/38720)

Restarting both

**Comment by kolyshkin** [comment]
derek add label: rebuild/experimental

**Comment by thaJeztah** [comment]
Was the last commit only moving the test to the new test-suite, or were changes made as well related to this PR?

Perhaps it's better to move the general test-changes to a separate PR

**Comment by kolyshkin** [comment]
> Was the last commit only moving the test to the new test-suite, or were changes made as well related to this PR?

TL;DR: this change is needed here and can't be separated out (unless we merge it first).

Perhaps I was too brief in my commit description:

> Move the test case from integration-cli to integration.
> This makes it use the current API version.

The test logic itself has not changed, except these two things:
* the new test sets default-ipc-mode via command line rather than via daemon.json (it doesn't matter);
* the new test uses current API version.

The problem this move solves is, old integration-cli tests use old API version, and so the following code stopped working (see https://github.com/moby/moby/pull/38700#issuecomment-462189895) after the changes in this PR:

1. start a daemon with "private" default-ipc-mode;
2. create a container without explicitly specifying its ipc mode;
3. check its mode is shareable.

When old API is used, check in (3) fails.

@thaJeztah I can split the last commit to a separate PR if you want

**Comment by thaJeztah** [comment]
My thinking was; do the general modernization of `TestDaemonRestartIpcMode` in a separate PR, and (in this PR) add the new test-case to the test

**Comment by kolyshkin** [comment]
>  do the general modernization of TestDaemonRestartIpcMode in a separate PR, and (in this PR) add the new test-case to the test

Done, test move separated out to https://github.com/moby/moby/pull/38843, so this one is simple to review. The only catch is #38843 need to be merged first for this PR to pass CI.

**Comment by kolyshkin** [comment]
Closing this in favor of last version of https://github.com/moby/moby/pull/35621
